### PR TITLE
API update

### PR DESCRIPTION
**Pull request by pnmadelaine**
This PR updates the API of various functions in HACL*.
See #675 and https://github.com/cryspen/hacl-packages/issues/255.
To merge with https://github.com/cryspen/hacl-packages/pull/377, https://github.com/hacl-star/merkle-tree/pull/13 and https://github.com/project-everest/mitls-fstar/pull/259.

## Blake2b

| Current                                                            | New                                                                                         |
|--------------------------------------------------------------------|---------------------------------------------------------------------------------------------|
| Hacl_Blake2b_32_blake2b(nn, output, ll, d, kk, k)                  | Hacl_Hash_Blake2b_hash_with_key(output, output_len, input, input_len, key, key_len)         |
| Hacl_Streaming_Blake2_blake2b_32_state                             | Hacl_Hash_Blake2b_state_t                                                                   |
| Hacl_Streaming_Blake2_blake2b_32_no_key_create_in(void)            | Hacl_Hash_Blake2b_malloc(void)                                                              |
| Hacl_Streaming_Blake2_blake2b_32_no_key_update(p, data, len)       | Hacl_Hash_Blake2b_update(state, chunk, chunk_len)                                           |
| Hacl_Streaming_Blake2_blake2b_32_no_key_finish(p, dst)             | Hacl_Hash_Blake2b_digest(state, output)                                                     |
| Hacl_Streaming_Blake2_blake2b_32_no_key_init(s1)                   | Hacl_Hash_Blake2b_reset(state)                                                              |
| Hacl_Streaming_Blake2_blake2b_32_no_key_free(s1)                   | Hacl_Hash_Blake2b_free(state)                                                               |
| Hacl_Blake2b_256_blake2b(nn, output, ll, d, kk, k)                 | Hacl_Hash_Blake2b_Simd256_hash_with_key(output, output_len, input, input_len, key, key_len) |
| Hacl_Streaming_Blake2b_256_blake2b_256_state                       | Hacl_Hash_Blake2b_Simd256_state_t                                                           |
| Hacl_Streaming_Blake2b_256_blake2b_256_no_key_create_in(void)      | Hacl_Hash_Blake2b_Simd256_malloc(void)                                                      |
| Hacl_Streaming_Blake2b_256_blake2b_256_no_key_update(p, data, len) | Hacl_Hash_Blake2b_Simd256_update(state, chunk, chunk_len)                                   |
| Hacl_Streaming_Blake2b_256_blake2b_256_no_key_finish(p, dst)       | Hacl_Hash_Blake2b_Simd256_digest(state, output)                                             |
| Hacl_Streaming_Blake2b_256_blake2b_256_no_key_init(s)              | Hacl_Hash_Blake2b_Simd256_reset(state)                                                      |
| Hacl_Streaming_Blake2b_256_blake2b_256_no_key_free(s)              | Hacl_Hash_Blake2b_Simd256_free(state)                                                       |

## Blake2s

| Current                                                            | New                                                                                         |
|--------------------------------------------------------------------|---------------------------------------------------------------------------------------------|
| Hacl_Blake2s_32_blake2s(nn, output, ll, d, kk, k)                  | Hacl_Hash_Blake2s_hash_with_key(output, output_len, input, input_len, key, key_len)         |
| Hacl_Streaming_Blake2_blake2s_32_state                             | Hacl_Hash_Blake2s_state_t                                                                   |
| Hacl_Streaming_Blake2_blake2s_32_no_key_create_in(void)            | Hacl_Hash_Blake2s_malloc(void)                                                              |
| Hacl_Streaming_Blake2_blake2s_32_no_key_update(p, data, len)       | Hacl_Hash_Blake2s_update(state, chunk, chunk_len)                                           |
| Hacl_Streaming_Blake2_blake2s_32_no_key_finish(p, dst)             | Hacl_Hash_Blake2s_digest(state, output)                                                     |
| Hacl_Streaming_Blake2_blake2s_32_no_key_init(s1)                   | Hacl_Hash_Blake2s_reset(state)                                                              |
| Hacl_Streaming_Blake2_blake2s_32_no_key_free(s1)                   | Hacl_Hash_Blake2s_free(state)                                                               |
| Hacl_Blake2s_128_blake2s(nn, output, ll, d, kk, k)                 | Hacl_Hash_Blake2s_Simd128_hash_with_key(output, output_len, input, input_len, key, key_len) |
| Hacl_Streaming_Blake2s_128_blake2s_128_state                       | Hacl_Hash_Blake2s_Simd128_state_t                                                           |
| Hacl_Streaming_Blake2s_128_blake2s_128_no_key_create_in(void)      | Hacl_Hash_Blake2s_Simd128_malloc(void)                                                      |
| Hacl_Streaming_Blake2s_128_blake2s_128_no_key_update(p, data, len) | Hacl_Hash_Blake2s_Simd128_update(state, chunk, chunk_len)                                   |
| Hacl_Streaming_Blake2s_128_blake2s_128_no_key_finish(p, dst)       | Hacl_Hash_Blake2s_Simd128_digest(state, output)                                             |
| Hacl_Streaming_Blake2s_128_blake2s_128_no_key_init(s)              | Hacl_Hash_Blake2s_Simd128_reset(state)                                                      |
| Hacl_Streaming_Blake2s_128_blake2s_128_no_key_free(s)              | Hacl_Hash_Blake2s_Simd128_free(state)                                                       |

## MD5

| Current                                               | New                                           |
|-------------------------------------------------------|-----------------------------------------------|
| Hacl_Streaming_MD5_legacy_hash(input, input_len, dst) | Hacl_Hash_MD5_hash(output, input, input_len)  |
| Hacl_Streaming_MD_state_32                            | Hacl_Hash_MD5_state_t                         |
| Hacl_Streaming_MD5_legacy_create_in(void)             | Hacl_Hash_MD5_malloc(void)                    |
| Hacl_Streaming_MD5_legacy_update(p, data, len)        | Hacl_Hash_MD5_update(state, chunk, chunk_len) |
| Hacl_Streaming_MD5_legacy_finish(p, dst)              | Hacl_Hash_MD5_digest(state, output)           |
| Hacl_Streaming_MD5_legacy_init(s)                     | Hacl_Hash_MD5_reset(state)                    |
| Hacl_Streaming_MD5_legacy_free(s)                     | Hacl_Hash_MD5_free(state)                     |
| Hacl_Streaming_MD5_legacy_copy(s0)                    | Hacl_Hash_MD5_copy(state)                     |

## SHA-1

| Current                                                | New                                            |
|--------------------------------------------------------|------------------------------------------------|
| Hacl_Streaming_SHA1_legacy_hash(input, input_len, dst) | Hacl_Hash_SHA1_hash(output, input, input_len)  |
| Hacl_Streaming_SHA1_state                              | Hacl_Hash_SHA1_state_t                         |
| Hacl_Streaming_SHA1_legacy_create_in(void)             | Hacl_Hash_SHA1_malloc(void)                    |
| Hacl_Streaming_SHA1_legacy_update(p, data, len)        | Hacl_Hash_SHA1_update(state, chunk, chunk_len) |
| Hacl_Streaming_SHA1_legacy_finish(p, dst)              | Hacl_Hash_SHA1_digest(state, output)           |
| Hacl_Streaming_SHA1_legacy_init(s)                     | Hacl_Hash_SHA1_reset(state)                    |
| Hacl_Streaming_SHA1_legacy_free(s)                     | Hacl_Hash_SHA1_free(state)                     |
| Hacl_Streaming_SHA1_legacy_copy(s0)                    | Hacl_Hash_SHA1_copy(state)                     |

## SHA-2

| Current                                             | New                                               |
|-----------------------------------------------------|---------------------------------------------------|
| Hacl_Streaming_SHA2_state_sha2_{n}                  | Hacl_Hash_SHA2_state_t_{n}                        |
| Hacl_Streaming_SHA2_create_in_{n}(void)             | Hacl_Hash_SHA2_malloc_{n}(void)                   |
| Hacl_Streaming_SHA2_init_{n}(s)                     | Hacl_Hash_SHA2_reset_{n}(state)                   |
| Hacl_Streaming_SHA2_finish_{n}(p, dst)              | Hacl_Hash_SHA2_digest_{n}(state, output)          |
| Hacl_Streaming_SHA2_free_{n}(p)                     | Hacl_Hash_SHA2_free_{n}(state)                    |
| Hacl_Streaming_SHA2_hash_{n}(input, input_len, dst) | Hacl_Hash_SHA2_hash_{n}(output, input, input_len) |
| Hacl_Streaming_SHA2_copy_{n}(s0)                    | Hacl_Hash_SHA2_copy_{n}(state)                    |
 
## SHA-3

| Current                                         | New                                               |
|-------------------------------------------------|---------------------------------------------------|
| Hacl_SHA3_sha3_{n}(inputByteLen, input, output) | Hacl_Hash_SHA3_sha3_{n}(output, input, input_len) |
| Hacl_Streaming_Keccak_sate                      | Hacl_Hash_SHA3_sate_t                             |
| Hacl_Streaming_Keccak_malloc(alg)               | Hacl_Hash_SHA3_malloc(alg)                        |
| Hacl_Streaming_Keccak_reset(s)                  | Hacl_Hash_SHA3_reset(state)                       |
| Hacl_Streaming_Keccak_finish(p, dst)            | Hacl_Hash_SHA3_digest(state, output)              |
| Hacl_Streaming_Keccak_free(s)                   | Hacl_Hash_SHA3_free(state)                        |
| Hacl_Streaming_Keccak_copy(s0)                  | Hacl_Hash_SHA3_copy(state)                        |

## Poly1305

| Current                                             | New                                                          |
|-----------------------------------------------------|--------------------------------------------------------------|
| Hacl_Poly1305_32_poly1305_mac(tag, len, text, key)  | Hacl_MAC_Poly1305_mac(output, input, input_len, key)         |
| Hacl_Streaming_Poly1305_32_poly1305_32_state        | Hacl_MAC_Poly1305_state_t                                    |
| Hacl_Streaming_Poly1305_32_create_in(k)             | Hacl_MAC_Poly1305_malloc(key)                                |
| Hacl_Streaming_Poly1305_32_init(k, s)               | Hacl_MAC_Poly1305_reset(state, key)                          |
| Hacl_Streaming_Poly1305_32_finish(p, dst)           | Hacl_MAC_Poly1305_digest(state, output)                      |
| Hacl_Streaming_Poly1305_32_update(p, data, len)     | Hacl_MAC_Poly1305_update(state, chunk, chunk_len)            |
| Hacl_Streaming_Poly1305_32_free(s)                  | Hacl_MAC_Poly1305_free(state)                                |
| Hacl_Poly1305_{n}_poly1305_mac(tag, len, text, key) | Hacl_MAC_Poly1305_Simd{n}_mac(output, input, input_len, key) |
| Hacl_Streaming_Poly1305_{n}_poly1305_{n}_state      | Hacl_MAC_Poly1305_Simd{n}_state_t                            |
| Hacl_Streaming_Poly1305_{n}_create_in(k)            | Hacl_MAC_Poly1305_Simd{n}_malloc(key)                        |
| Hacl_Streaming_Poly1305_{n}_init(k, s)              | Hacl_MAC_Poly1305_Simd{n}_reset(state, key)                  |
| Hacl_Streaming_Poly1305_{n}_update(p, data, len)    | Hacl_MAC_Poly1305_Simd{n}_update(state, chunk, chunk_len)    |
| Hacl_Streaming_Poly1305_{n}_finish(p, dst)          | Hacl_MAC_Poly1305_Simd{n}_digest(state, output)              |
| Hacl_Streaming_Poly1305_{n}_free(s)                 | Hacl_MAC_Poly1305_Simd{n}_free(state)                        |

## EverCrypt

| Current                                         | New                                                        |
|-------------------------------------------------|------------------------------------------------------------|
| EverCrypt_Hash_Incremental_hash_state           | EverCrypt_Hash_Incremental_state_t                         |
| EverCrypt_Hash_Incremental_create_in(a)         | EverCrypt_Hash_Incremental_malloc(a)                       |
| EverCrypt_Hash_Incremental_init(s)              | EverCrypt_Hash_Incremental_reset(state)                    |
| EverCrypt_Hash_Incremental_update(s, data, len) | EverCrypt_Hash_Incremental_update(state, chunk, chunk_len) |
| EverCrypt_Hash_Incremental_finish(s, dst)       | EverCrypt_Hash_Incremental_digest(state, output)           |
| EverCrypt_Hash_Incremental_free(s)              | EverCrypt_Hash_Incremental_free(state)                     |
| EverCrypt_Poly1305_poly1305(dst, src, len, key) | EverCrypt_Poly1305_mac(output, input, input_len, key)      |

## Headers

| Current                       | New                                      |
|-------------------------------|------------------------------------------|
| Hacl_Hash_Blake2.h            | Hacl_Hash_Blake2b.h, Hacl_Hash_Blake2s.h |
| Hacl_Streaming_Blake2b.h      | Hacl_Hash_Blake2b.h                      |
| Hacl_Streaming_Blake2s.h      | Hacl_Hash_Blake2s.h                      |
| Hacl_Hash_Blake2b_256.h       | Hacl_Hash_Blake2b_Simd256.h              |
| Hacl_Hash_Blake2s_128.h       | Hacl_Hash_Blake2s_Simd128.h              |
| Hacl_Streaming_Blake2b_256.h  | Hacl_Hash_Blake2b_Simd256.h              |
| Hacl_Streaming_Blake2s_128.h  | Hacl_Hash_Blake2s_Simd128.h              |
| Hacl_Poly1305_32.h            | Hacl_MAC_Poly1305.h                      |
| Hacl_Streaming_Poly1305_32.h  | Hacl_MAC_Poly1305.h                      |
| Hacl_Poly1305_128.h           | Hacl_MAC_Poly1305_Simd128.h              |
| Hacl_Streaming_Poly1305_128.h | Hacl_MAC_Poly1305_Simd128.h              |
| Hacl_Poly1305_256.h           | Hacl_MAC_Poly1305_Simd256.h              |
| Hacl_Streaming_Poly1305_256.h | Hacl_MAC_Poly1305_Simd256.h              |

### CODE REVIEW

File: dist/gcc-compatible/EverCrypt_Hash.h
Comment (franziskuskiefer) [inline dist/gcc-compatible/EverCrypt_Hash.h:120]: Shouldn't the `output` come first?

File: dist/gcc-compatible/Hacl_Chacha20Poly1305_128.h
Comment (franziskuskiefer) [inline dist/gcc-compatible/Hacl_Chacha20Poly1305_128.h:1]: I presume these files get renamed in a different PR because this one is only about hashing.

File: dist/gcc-compatible/Hacl_HMAC_Blake2b_256.h
Comment (franziskuskiefer) [inline dist/gcc-compatible/Hacl_HMAC_Blake2b_256.h:1]: Should these files be renamed (`_Simd256`)?

File: dist/gcc-compatible/Hacl_Hash_Blake2b.h
Comment (franziskuskiefer) [inline dist/gcc-compatible/Hacl_Hash_Blake2b.h:44]: These functions need documentation.

File: dist/gcc-compatible/Hacl_Hash_Blake2b.h
Comment (franziskuskiefer) [inline dist/gcc-compatible/Hacl_Hash_Blake2b.h:45]: I'm not sure what `wv` is. But the arguments in these stateful functions appear to be in the wrong order.

File: dist/gcc-compatible/Hacl_Hash_Blake2s.h
Comment (franziskuskiefer) [inline dist/gcc-compatible/Hacl_Hash_Blake2s.h:43]: Same here. Can we get comments?

File: dist/gcc-compatible/Hacl_Hash_Blake2s_Simd128.h
Comment (franziskuskiefer) [inline dist/gcc-compatible/Hacl_Hash_Blake2s_Simd128.h:45]: Comments.

File: dist/gcc-compatible/Hacl_Hash_SHA2.h
Comment (franziskuskiefer) [inline dist/gcc-compatible/Hacl_Hash_SHA2.h:80]: Al the SHA/MD functions are lacking comments. But we can do this in a follow up.

File: dist/gcc-compatible/Hacl_Hash_SHA3.h
Comment (franziskuskiefer) [inline dist/gcc-compatible/Hacl_Hash_SHA3.h:42]: What does the `256` in the name supposed to convey here?

File: dist/gcc-compatible/Hacl_MAC_Poly1305.h
Comment (franziskuskiefer) [inline dist/gcc-compatible/Hacl_MAC_Poly1305.h:46]: Wasn't the length arugment supposed to come ofter the data? Or what's the `len` here?

**Review by franziskuskiefer — CHANGES_REQUESTED** [review]
Thanks! 🎉 finally getting better names/APIs. I left some comments.

We should make sure we don't conflict/do unnecessary work with #756. @msprotz can you make sure to look at this?

File: dist/gcc-compatible/EverCrypt_Hash.h
Comment (pnmadelaine) [reply dist/gcc-compatible/EverCrypt_Hash.h:120]: I treated the "algorithm" input as a special input, like "state" in the streaming API, thus letting it at the beginning

File: dist/gcc-compatible/Hacl_HMAC_Blake2b_256.h
Comment (pnmadelaine) [reply dist/gcc-compatible/Hacl_HMAC_Blake2b_256.h:1]: Yes, but I thought we would tackle the HMAC modules in a separate PR.

File: dist/gcc-compatible/EverCrypt_Hash.h
Comment (franziskuskiefer) [reply dist/gcc-compatible/EverCrypt_Hash.h:120]: The state is special because it's really an i/o parameter. This is a regular input imo. Has there been a discussion for leaving it in front here? Not sure what others think (@msprotz?).

Comment (franziskuskiefer) [reply dist/gcc-compatible/EverCrypt_Hash.h:120]: The state is special because it's really an i/o parameter. This is a regular input imo. Has there been a discussion for leaving it in front here? Not sure what others think (@msprotz?).

File: dist/gcc-compatible/Hacl_Hash_Blake2b.h
Comment (franziskuskiefer) [reply dist/gcc-compatible/Hacl_Hash_Blake2b.h:44]: They are here, so they are in the public API 😉

File: dist/gcc-compatible/Hacl_Hash_SHA3.h
Comment (pnmadelaine) [reply dist/gcc-compatible/Hacl_Hash_SHA3.h:42]: This comes from the current names.
Since they were discussions on SHA2 and SHA3 being updated soon we agreed to update their API afterwards.
The only modifications to this file are caused by the streaming functor update.

File: dist/gcc-compatible/Hacl_Hash_SHA3.h
Comment (pnmadelaine) [reply dist/gcc-compatible/Hacl_Hash_SHA3.h:58]: As mentioned above, the SHA2 and SHA3 were only marginally modified, the rest of the work is left to a future PR.

**Comment by pnmadelaine** [comment]
As a global response to some of the comments, I should point out that:

- SHA2 and SHA3 are left pretty much as is because there was a discussion about waiting for an update in the near future before touching their API. The only modifications to these modules are caused by the streaming functor update.

- I did not touch some functions that appear in the header files, because either I got confused and it seemed to me they were not supposed to be called by the user anyway (like some init functions), or they were redundant (like the streaming API for poly1305 being defined twice). As a general rule I only touched declarations that appeared in the hacl-packages documentation.

**Comment by msprotz** [comment]
> We should make sure we don't conflict/do unnecessary work with #756. @msprotz can you make sure to look at this?

That's fine. I'm behind on that PR anyhow (working on and off on it). I don't mind merging master in.

File: dist/gcc-compatible/EverCrypt_Hash.h
Comment (msprotz) [reply dist/gcc-compatible/EverCrypt_Hash.h:120]: honestly I don't really have much of an opinion here. I think agile APIs take the agility parameter first because of the original type in F*, of the form:

```
let hash_t (a: alg) = output:hash a -> input:buffer uint8 -> len: uint32 { len == length input }
val hash #a: hash_t a
let hash #a output input len = ...
```

so you might have a hard time changing that without making signature very difficult to deal with in the source (we can, but is it worth it?)

File: dist/gcc-compatible/Hacl_Hash_Blake2b.h
Comment (msprotz) [reply dist/gcc-compatible/Hacl_Hash_Blake2b.h:45]: this is another thing that is caused by the original F* type, which uses the Lib helpers -- this could be fixed at the expense of defining other abbreviations, such as `length_of` instead of `lbuffer` -- hopefully this won't break the verification

FYI, `wv` is working vector, but it took me years to understand that... I think we should also fix argument order and names while we're at it

File: dist/gcc-compatible/Hacl_MAC_Poly1305.h
Comment (msprotz) [reply dist/gcc-compatible/Hacl_MAC_Poly1305.h:46]: yeah same issue the order of arguments is incorrect

File: dist/gcc-compatible/EverCrypt_HMAC.h
Comment (msprotz) [inline dist/gcc-compatible/EverCrypt_HMAC.h:48]: this is quite unsightly but I guess this isn't something we're focusing on here

**Review by msprotz — COMMENTED** [review]
This is a great PR -- congratulations!

My main comment is essentially: how much should be in scope for this PR. I'll reiterate my earlier point: if we break the API, better do it while the functions are changing names, which will force callers and users to figure things out.

Here's my personal take on what is and isn't in scope.
- position of the algorithm parameter: not a fight worth picking, it's going to generate a lot of strife on the HACL* side and would delay this PR by a lot
- input, len vs len, input: must fix this right now, this is our only shot at it, the blake2 functions are changing files, so we might as well fix their APIs; pnm, let me know if you need help defining a `Lib.Buffer.length_of` type abbreviation to help with that
- too many functions are visible in the headers: this should be the next PR after this one, but isn't strictly essential
- sha2/sha3: forthcoming, let's also not bother about it
- renaming arguments (wv, ww, vv, etc.) also a followup, not an ABI-breaking change

Please let me know your thoughts

File: dist/gcc-compatible/Hacl_Hash_Blake2b.h
Comment (pnmadelaine) [reply dist/gcc-compatible/Hacl_Hash_Blake2b.h:44]: `Hacl_Hash_Blake2b_init` is used in `Hacl_Hash_Blake2b_hash_with_key` and the `init` function of `EverCrypt_Hash` as well as in `Hacl_HMAC_compute_blake2b_32`.

`Hacl_Hash_Blake2b_update_key` is only used in `Hacl_Hash_Blake2b.c`.

File: dist/gcc-compatible/Hacl_MAC_Poly1305.h
Comment (pnmadelaine) [reply dist/gcc-compatible/Hacl_MAC_Poly1305.h:46]: But what do these functions do? `Hacl_MAC_Poly1305_poly1305_update1` isn't used anywhere and `Hacl_MAC_Poly1305_poly1305_update` seems to me like an auxiliary function used in `Hacl_MAC_Poly1305_mac` and `Hacl_MAC_Poly1305_update`.
In both cases, the logical name for them would be `Hacl_MAC_Poly1305_update` which is redundant with the streaming API.

File: dist/gcc-compatible/Hacl_Hash_Blake2s.h
Comment (pnmadelaine) [reply dist/gcc-compatible/Hacl_Hash_Blake2s.h:51]: Where is `Hacl_Hash_Blake2s` marked as `noextract`? `update_key` does not seem to be used anywhere, the others seem to be used for instance in `Hacl_HMAC` or in `EverCrypt_Hash`.

File: dist/gcc-compatible/Hacl_MAC_Poly1305.h
Comment (msprotz) [reply dist/gcc-compatible/Hacl_MAC_Poly1305.h:46]: ah! I confused those with the streaming API. I think this module should undergo the same treatment: hide Hacl.Poly1305 underneath Hacl.Streaming.Poly1305

**Review by msprotz — APPROVED** [review]
With the most recent commit, I am quite happy with the changes. All of the problematic API entry points are hidden, so now a lot of discussions are moot.

Final question, does this need to happen for Poly1305 too?

**Review by franziskuskiefer — APPROVED** [review]
Thanks for getting rid of those functions! I'd fine getting this in as is now.

**Comment by pnmadelaine** [comment]
> Final question, does this need to happen for Poly1305 too?

@msprotz imo it should

**Comment by pnmadelaine** [comment]
also, I'd really like the unstable proof in `Hacl.Streaming.Functor` to be fixed before merging. @msprotz @R1kM could you take a look?

**Comment by R1kM** [comment]
@pnmadelaine Is the unstable proof the one where the rlimit was bumped from 100 to 500? Given the complexity of the code, 500 does not look too crazy to me, and the proof is already quite detailed. I'd be happy to take it as is.

**Comment by msprotz** [comment]
One last point I would like to bring up... apologies for raising this so late in the process, but this is great work and I want to make sure the landing of this PR goes smoothly.

My concern is about clients of hacl-star who may have to update their code in order to "catch up" with the renamings. We know from experience that this is something that is almost always poorly received: this generates extra work on our clients and champions who love HACL*, and so they're likely to be annoyed at having to do extra work for no good reason (from their point of view).

Franziskus and I had a brief discussion about it this morning, pnm and I touched on the topic briefly on Slack. Here are some points that came up.
- one may need to define macros for backwards-compat
- static inline wrappers are better than macros because they can be marked with KRML_DEPRECATED (and we have a mechanism to do that already from F*)
- some functions have *not* been renamed (e.g. update), and as such, cannot receive a deprecation warning, unless we start having `update` and `update2`, but this might be a little extreme
- this may not be a big deal in any case: few of our clients use the C API, right now, it's only Mozilla and Python, and we could conceivably submit PRs on both sides to make sure they remain up-to-date

Would love to have @pnmadelaine's, @franziskuskiefer and possibly @karthikbhargavan's thoughts. Thanks!

**Comment by franziskuskiefer** [comment]
> My concern is about clients of hacl-star who may have to update their code in order to "catch up" with the renamings.

Adding a deprecation warning and keeping the old functions is nice to have I think. But given that we have the argument reordering without renaming as well, this might actually introduce more confusion.
I'm a little on the fence on this. On the one hand it's nice not to fail hard for anyone just updating. On the other hand there's no stable release that would suggest that the APIs are stable.

With this in mind I'd suggest to add deprecation warnings where we can if it's not too much work, but not spend too much time/block too long on this.

**Comment by msprotz** [comment]
We had a long discussion about this in the maintainers' meeting this morning. This is really essential, important work, and we want to make sure that this lands successfully. Thanks again to Paul-Nicolas and Damian for leading this effort. The following points were brought up.
- It is essential to make this happen in one go. We don't want consumers of HACL to chase API changes every few weeks while we land various PRs. So, let's try to make *all* of the renamings happen here, not just hashes.
- This means that we need to make the SHA2 and SHA3 work happen urgently. SHA3 is ready to merge (pending review). Aymeric and I will look at SHA2 tomorrow.
- Submitting patches on our own to Mozilla & Python to make them pick up the latest changes is a good idea.
- Let's make sure we keep this PR up to date so that it doesn't bit-rot.
- I am sending an email to the community (right now, hacl contributors, since we don't have a mailing-list for announcements) to give everyone visibility about this change.

**Comment by msprotz** [comment]
Question for @pnmadelaine: what other renamings do you have on your radar? Can you and I talk about the other renamings that need to happen? How can I best help with the rest of the work?

Thanks again for driving this effort! This is a much-needed and much-appreciated change.

**Review by msprotz — APPROVED** [review]
Aymeric and I looked at this -- thanks for this super involved piece of work, looks good!

Can you please submit a followup PR that adds an entry in Changes.md that describes the changes, possibly reusing the table from this PR?

**Pull request merged**
